x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderService.java

```diff
@@ -101,9 +102,21 @@ public List<DiscoveryNodeRole> roles() {
 
     @Override
     public AutoscalingDeciderResult scale(Settings configuration, AutoscalingDeciderContext context) {
+        Set<ShardId> unassignedShardIds = StreamSupport.stream(context.state().getRoutingNodes().unassigned().spliterator(), false)
+            .map(ShardRouting::shardId)
+            .collect(Collectors.toSet());
+        Set<ShardId> assignedShardIds = context.state()
```

Comment:
The meaning of "assigned" in this decider is somewhat different from this. It means the assigned shards that contributes to the autoscaling decision. It is shards that are assigned to a node, where `canRemain() == NO` and there is no other node where it can be assigned. Additionally, either the reason it cannot reamin or it cannot be allocated must be a single disk threshold reason (i.e., not for other reasons).

I think you will need to get this information out together with the amount of disk space needed, i.e., to change `storagePreventsRemainOrMove` to return both the number of the list of shards.